tests/src/tests.rs

```diff
@@ -570,73 +599,216 @@ fn test_part(
-            write!(output, "    Not annotated | ").unwrap();
-            print_annotation(output, &source, line, unexpected)
+            for missing in missing_outputs {
+                write!(output, "    Not emitted   | ").unwrap();
+                print_annotation(output, &source, line, missing)
+            }
+        }
+    } else {
+        // warns and ignores diagnostics
+        if !diagnostic_annotations.is_empty() {
+            writeln!(
+                output,
+                "  Subtest {i} contains diagnostics but is in autocomplete mode."
+            )
+            .unwrap();
+            for a in diagnostic_annotations {
+                write!(output, "    Ignored | ").unwrap();
+                print_annotation(output, &source, line, &a);
+            }
         }
 
-        for missing in missing_outputs {
-            write!(output, "    Not emitted   | ").unwrap();
-            print_annotation(output, &source, line, missing)
+        for annotation in metadata.annotations.iter().filter(|a| {
+            matches!(
+                a.kind,
+                AnnotationKind::AutocompleteContains
+                    | AnnotationKind::AutocompleteExcludes
+            )
+        }) {
+            let cursor = if let Some(range) = &annotation.range {
+                if range.start != range.end {
+                    writeln!(
+                        output,
+                        "  Subtest {i} has this annotation using a range where `range.len() != 0`."
+                    )
+                    .unwrap();
+                    writeln!(
+                        output,
+                        "  `range.start` will be used and `range.end` will be ignored."
+                    )
+                    .unwrap();
+                    write!(output, "    Ignored r.end | ").unwrap();
+                    print_annotation(output, &source, line, annotation);
+                }
+                range.start
+            } else {
+                writeln!(
+                    output,
+                    "  Subtest {i} has an autocomplete annotation but no range specified"
+                )
+                .unwrap();
+                write!(output, "Annotation ignored| ").unwrap();
+                print_annotation(output, &source, line, annotation);
+                ok = false;
+                continue;
+            };
+            // todo, use document if is_some to test labels autocomplete
+            let completions = typst_ide::autocomplete(world, None, &source, cursor, true)
+                .map(|(_, c)| c)
+                .unwrap_or_default()
+                .into_iter()
+                .map(|c| c.label.to_string())
+                .collect::<HashSet<_>>();
+            let completions =
+                completions.iter().map(|s| s.as_str()).collect::<HashSet<&str>>();
+
+            let must_contain_or_exclude = parse_autocomplete_message(&annotation.message);
+            let missing =
```

Comment:
could live in the next block but then we'd need another if statement inside to check  if it's empty, doesn't seem elegant or worth it